src/ipc/ipc3/handler.c

```diff
@@ -254,7 +254,10 @@ static int ipc_stream_pcm_params(uint32_t stream)
 		 */
 		pipe_dbg(pcm_dev->cd->pipeline,
 			 "ipc: ignore PCM param change request on an active pipeline");
-		return 0;
+		/*To resolve the HS playback not happening when
+		 *speaker is played first in tdm use case.
+		 */
+		break;
```

Comment:
how do we know this is for a tdm use-case? For other usecases this isnt valid. And BTW this whole switch-case is useless with your change

Reply:
Could you please take a look at the topology change also, it seems it is relevant based on the topology.
If it is breaking some other usecase or platform, can we use a macro.

Reply:
@maruthi-machani I think needs some solution that works more generally. Why is PCM params called multiple times in your configuration, can you explain a bit the use-case and sequence? Maybe we can figure out how this could be handled more generally based on that.

Reply:
![tplg_mux](https://user-images.githubusercontent.com/101189005/200791404-50ee69a9-c52b-4744-ae29-b110e8507ecb.png)

usecase: Simultaneous playback of two different streams on headset and speaker using single I2S controller.
                when a stream is playing on speaker (pipe1 w.r.t attached topology graph) and  if we start another stream on headset (pipe0) then the headset stream is not playing. pipe0 is returning from the above **"return 0"** code.
(issue not observed with headset first and speaker later usecase)

Attached the topology graph for reference.

Reply:
@maruthi-machani Thanks for the use-case info. I'll unresolve the discussion, so other reviewers can evaluate this. 

I think this could be a limitation and/or a bug. The check was added by this commit:
```
commit 5ba541577701943e21ba79e4511feb85b7ec7e98
Author: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>
Date:   Tue Sep 28 13:57:28 2021 +0200

    ipc: ipc3: ignore PCM_PARAMS for active pipelines
```

... It would seem here we still want to skip the pipeline setup (as it is already setup up), but we need to still do some steps for pcm0 in the function.

Reply:
@maruthi-machani What is the comp_id in the PCM params IPC payload when you start the 2 PCM's?

Reply:
I dont think we can share a single buffer to DAIs - I think we may need a buffer for each ACPSP. This may also cause an issue for HW DMA ? i.e. 2 channels trying to read same buffer at same time ?

Reply:
As mentioned in above comment, ACPSP1 is a virtual dai. There is no data access from ACPSP1.

Reply:
Just curious why it needs to be virtual vs just using a single dai?

Reply:
@maruthi-machani even if we want to proceed with this change, I will say that the switch/case block is absolutely redundant and the debug message is incorrect after the change. Please fix it

Reply:
> Just curious why it needs to be virtual vs just using a single dai?

Driver expects the DAI to bind the PCM and send H/W params IPC so that two end points are exposed to OS, which allows the simultaneous playback. Hence virtual DAI is required.

Reply:
> I dont think we can share a single buffer to DAIs - I think we may need a buffer for each ACPSP. This may also cause an issue for HW DMA ? i.e. 2 channels trying to read same buffer at same time ?

It is 4 channels in TDM mode with 2 channels each for ACPSP and ACPSP1.

Reply:
It was my commit https://github.com/thesofproject/sof/pull/4869/commits/c3e6d86aaee01b4f6ef38fb3dd8a6b411ea5703c that introduced that `return 0` more than a year ago, so I don't remember any more if we actually have any such flows that hit that path. The proposed change is equivalent to just removing the whole `switch` statement completely. Instead, to stay conservative, I'd make this change conditional on your specific DAI type, if it's really needed after the questions with the topology are resolved.

Reply:
> @maruthi-machani I don't think that kind of topology is supported, buffers shouldn't have multiple sink links. You probably need a demux there

@lyakh could you please take a look at the comments in below link, the requirement is to have a simultaneous playback on a single I2S, using tdm mode. Also demux doesn't help here, because ACPSP1 is a virtual DAI.
https://github.com/thesofproject/sof/pull/6559#issuecomment-1313431423

Reply:
> > @maruthi-machani I don't think that kind of topology is supported, buffers shouldn't have multiple sink links. You probably need a demux there
> 
> @lyakh could you please take a look at the comments in below link, the requirement is to have a simultaneous playback on a single I2S, using tdm mode. Also demux doesn't help here, because ACPSP1 is a virtual DAI. [#6559 (comment)](https://github.com/thesofproject/sof/pull/6559#issuecomment-1313431423)

Well, sorry, I don't have a good solution, but from what I know this topology isn't valid, so a different solution is needed. However, I'm also not sure that using a correct solution would immediately automatically solve your issue

Reply:
sorry, it is for IPC3, not IPC4.

Reply:
As suggested by @lgirdwood, with separate pipeline for mux->buf->DAI, the "return 0" issue occured.